# DUO.py

This notebook is the bases for the script_for_duos_playlist.py. It covers all steps which lead to the creation of a new duos playlist for my girlfriend and me.

The notebook was used for experimentation and building the steps to arrive at a DUO.py playlist that was enjoyable to both users.

I wanted to create a playlist that was different from Spotify's 'Duo Mix', which didn't really seem to grasp the music tastes of my girlfriend and me and was somehow disappointing.

The playlist we had hoped for wasn't available so I decided to create one, that could be updated periodically and would incorporate some features:
- include some all time favorites as well as songs we recently both enjoyed
- include new songs based on our favorite songs
- don't include too many songs from the same artist or genre
- provide a listening experience that is based on mutually liked music not just a combination of our individual tastes

Right now the workflow accomplishes this although it is not yet perfect. I intend to add machine learning to further "filter" songs based on what we would like to have in our personal playlist. This howoever requires some data that is currently collected as we enjoy and discard songs from the weekly playlist.

## Imports

The most important imports are spotipy, pandas, numpy and my own small library 'spotifuncs'. Here I use the wildcard import for experimentation, the scripts only imports necessary functions. spotifuncs itself uses pandas, sklearn, spotipy.

In [1]:
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
import spotipy.util as util
from pathlib import Path
import pandas as pd
import numpy as np
from spotifuncs import *

In [2]:
path = Path("C:/Users/ms101/OneDrive/DataScience_ML/projects/spotify_app")

## Setting the Credentials 

I stored the credentials and usernames in a .txt that was not uploaded to GitHub to ensure safety and avoid showing sensible information in my notebooks. 

Here I simply read the lines within the .txt files to retrieve the necessary information

In [3]:
with open(path / "client_s.txt") as f:
    content = f.readlines()
content = [x.strip() for x in content]

client_id = content[0]
client_secret = content[1]

In [4]:
with open(path / "usernames.txt") as f:
    usernames = f.readlines()
usernames = [x.strip() for x in usernames]

username1 = usernames[0]
username2 = usernames[1]

## App Scope

For the following code to work the scope needs to be quite extensive as I am retrieving a lot of user data and modifying a collaborative playlist (which the SpotifyAPI automatically views as private). As the only users are my girlfriend and me this was okay. 

**If others should choose to use my app a thorough explanation concerning the use of their data and the abilities of the app are paramount.**


To view all available scopes click here: https://developer.spotify.com/documentation/general/guides/scopes/

In [5]:
scope = "user-library-read user-read-recently-played user-top-read playlist-modify-public playlist-read-private playlist-read-collaborative playlist-modify-private"

redirect_uri = "https://developer.spotify.com/dashboard/applications/4a4e029d299a4241873db8300038bf0a"

client_credentials_manager = SpotifyClientCredentials(client_id=client_id, 
                                                      client_secret=client_secret)


## Authenticate

I reduced the authentication process to a function that can be found within the spotifuncs library. I describe it in more detail in my [Medium post](https://towardsdatascience.com/using-python-to-refine-your-spotify-recommendations-6dc08bcf408e).

In [6]:
sp_m = authenticate(redirect_uri, client_credentials_manager, username1, scope, client_id, client_secret)

In [8]:
sp_t = authenticate(redirect_uri, client_credentials_manager, username2, scope, client_id, client_secret)

## Retrieving the Data

I am retrieving quite a few dictonaries that end up in multiple dataframes. I packaged the process into a function because I have to repeat it. I decided not to integrate it into spotifuncs, as the function is very specific and subject to changes.

The first important piece of data are all the top tracks of the user, which I combine into a complete list of top tracks that cover everything from short to long term favorites. Notice that I am sampling only 15 songs from the long term top tracks and doing so **without** setting a random seed to **not** get the same results everytime the code is run. Always using the full list of long term favorites in the process would lead to too much repition over the course of multiple weeks and thus playlists.

The second important piece are the top artists, which like the tracks are retrieved for all time frames. The artists are important for the filtering process later on.

Lastly I am also retrieving the last 50 tracks a user saved. 50 is the upper limit here, which is unfortunate as this really limits the use of the data.

In [9]:
def get_dfs(sp):
    ##queries
    #user top tracks
    top_tracks_short = sp.current_user_top_tracks(limit = 50,offset=0,time_range='short_term')
    top_tracks_med = sp.current_user_top_tracks(limit = 50,offset=0,time_range='medium_term')
    top_tracks_long = sp.current_user_top_tracks(limit = 50,offset=0,time_range='long_term')
    
    #combine the top_tracks
    top_tracks_short_df = append_audio_features(create_df_top_songs(top_tracks_short),sp)
    top_tracks_med_df = append_audio_features(create_df_top_songs(top_tracks_med),sp)
    top_tracks_long_df = append_audio_features(create_df_top_songs(top_tracks_long),sp)
    #sample from long-term top tracks to introduce more randomness and avoid having the same artists
    top_tracks_long_df = top_tracks_long_df.sample(n = 15)
    top_tracks_df = pd.concat([top_tracks_short_df,top_tracks_med_df,top_tracks_long_df]).drop_duplicates().reset_index(drop = True)
        
    #user top artists
    top_artists_long = sp.current_user_top_artists(limit = 50, time_range = "long_term")
    top_artists_med = sp.current_user_top_artists(limit = 50, time_range = "medium_term")
    top_artists_short = sp.current_user_top_artists(limit = 50, time_range = "short_term")
    
    artists_short_df = top_artists_from_API(top_artists_short)
    artists_med_df = top_artists_from_API(top_artists_med)
    artists_long_df = top_artists_from_API(top_artists_long)
    artists_df = pd.concat([artists_short_df,artists_med_df,artists_long_df])
    artists_df["genres"] = artists_df["genres"].apply(lambda x: ",".join(x))
    artists_df.drop_duplicates().reset_index(drop = True)
    
    #user saved tracks
    user_saved_tracks = sp.current_user_saved_tracks(limit = 50)
    saved_tracks_df = create_df_saved_songs(user_saved_tracks)
    
        
    return top_tracks_df,artists_df,saved_tracks_df

In [10]:
top_tracks_m, artists_m, saved_tracks_m = get_dfs(sp_m)

In [11]:
top_tracks_t, artists_t, saved_tracks_t = get_dfs(sp_t)

In [12]:
artists_t

name                      id  \
0     Chelina  3XQZW9cuoDf7JhPbr99bXD   
1    Coldplay  4gzpq5DPGxSnKTe4SA8HAU   
2       Andro  4J6A7DGmVEA4CXhTnCxxEd   
3      Ghetts  7zJL978NtANOysfGY21ty6   
4         RIN  18ISxWwWjV6rPLoVCXf1dz   
..        ...                     ...   
45     Rakede  4soVkCNrRQccCv4Nohz273   
46     257ers  6ihLfpY3cmdGyWEnItn30w   
47      MEUTE  1z5xbcOeFRQXBVDpvRPh8H   
48  Max Herre  7IpWQKu80qQvyer3LO6SW3   
49     Hozier  2FXC3k01G6Gw61bmprjgqS   

                                               genres  popularity  \
0                                         amharic pop          14   
1                                  permanent wave,pop          89   
2                            russian pop,russian trap          61   
3             grime,uk alternative hip hop,uk hip hop          59   
4                     german cloud rap,german hip hop          72   
..                                                ...         ...   
45                                    hamburg hip hop          41   
46  antideutsche,deep german hip hop,german hip ho...          62   
47        german dance,hamburg electronic,livetronica          54   
48   german alternative rap,german hip hop,german pop          53   
49            irish singer-songwriter,modern rock,pop          81   

                                      uri  
0   spotify:artist:3XQZW9cuoDf7JhPbr99bXD  
1   spotify:artist:4gzpq5DPGxSnKTe4SA8HAU  
2   spotify:artist:4J6A7DGmVEA4CXhTnCxxEd  
3   spotify:artist:7zJL978NtANOysfGY21ty6  
4   spotify:artist:18ISxWwWjV6rPLoVCXf1dz  
..                                    ...  
45  spotify:artist:4soVkCNrRQccCv4Nohz273  
46  spotify:artist:6ihLfpY3cmdGyWEnItn30w  
47  spotify:artist:1z5xbcOeFRQXBVDpvRPh8H  
48  spotify:artist:7IpWQKu80qQvyer3LO6SW3  
49  spotify:artist:2FXC3k01G6Gw61bmprjgqS  

[108 rows x 5 columns]

In [13]:
artists_m

name                      id  \
0               badmómzjay  7oWrEQO1d3klp0Qrfh7a5h   
1                      RIN  18ISxWwWjV6rPLoVCXf1dz   
2                      SCH  2kXKa3aAFngGz2P4GjG5w2   
3                     SXTN  0tMFcqLXhtm1Gep20iuIR3   
4                   G-Eazy  02kJSzxNuaWGqwubyUba0Z   
..                     ...                     ...   
45              Chris Rock  36eSjIksD6fehqxyDUHDA3   
46                Wardruna  0NJ6wlOAsAJ1PN4VRdTPKA   
47                 Carnage  7CCjtD0hCK005Bvg2WG1a7   
48               Motörhead  1DFr97A9HnbV3SKTJFu62M   
49  Thirty Seconds To Mars  0RqtSIYZmd4fiBKVFqyIqD   

                                               genres  popularity  \
0               frauenrap,german drill,german hip hop          67   
1                     german cloud rap,german hip hop          72   
2   french hip hop,pop urbaine,rap francais,rap ma...          78   
3        deep german hip hop,frauenrap,german hip hop          61   
4   hip hop,indie pop rap,oakland hip hop,pop rap,rap          83   
..                                                ...         ...   
45                                black comedy,comedy          48   
46                  nordic folk,rune folk,viking folk          63   
47  edm,electro house,electronic trap,pop rap,trap...          64   
48  album rock,alternative metal,hard rock,metal,r...          69   
49              modern rock,pop punk,post-grunge,rock          71   

                                      uri  
0   spotify:artist:7oWrEQO1d3klp0Qrfh7a5h  
1   spotify:artist:18ISxWwWjV6rPLoVCXf1dz  
2   spotify:artist:2kXKa3aAFngGz2P4GjG5w2  
3   spotify:artist:0tMFcqLXhtm1Gep20iuIR3  
4   spotify:artist:02kJSzxNuaWGqwubyUba0Z  
..                                    ...  
45  spotify:artist:36eSjIksD6fehqxyDUHDA3  
46  spotify:artist:0NJ6wlOAsAJ1PN4VRdTPKA  
47  spotify:artist:7CCjtD0hCK005Bvg2WG1a7  
48  spotify:artist:1DFr97A9HnbV3SKTJFu62M  
49  spotify:artist:0RqtSIYZmd4fiBKVFqyIqD  

[112 rows x 5 columns]

### Finding common artists

I am finding common artists (of the 2 users) to later filter there top songs by artists. 
The logic behind this is the following:

A track might be only among one users "top tracks", it may however be by an artist both users enjoy. In that case the track is a good candidate for the duos playlist as both users will probably enjoy it, yet it might be a new discovery for one of them. If both already know and like it: still a good fit for the playlist!

In [14]:
common_artists = dataframe_difference(artists_m,artists_t, which = "both")
common_artists

C:\Users\ms101\Anaconda3\envs\spotify_app\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


name                      id  \
0              RIN  18ISxWwWjV6rPLoVCXf1dz   
1           G-Eazy  02kJSzxNuaWGqwubyUba0Z   
2       The Weeknd  1Xyo4u8uXC1ZmMpatF05PJ   
3             ELIF  65AzRSW0jKSs0WtttEXrOw   
4     Joyner Lucas  6C1ohJrd5VydigQtaGy5Wa   
5      KitschKrieg  5tHiL8SKSaZGMBUPIiSmX4   
6   Arctic Monkeys  7Ln80lUS6He07XvHI8qqHH   
7    Frank Sinatra  1Mxqyy3pSjf8kZZL4QVxS0   
8            Seeed  5ISjkNS17JpCwiFtW80lpV   
9             Juju  4sg4no0TXdsrM1s4SVUwNF   
10    Modeselektor  2jYMYP2SVifgmzNRQJx3SJ   
11          Eminem  7dGJo4pcD2V6oG8kP0tJRR   
12        Kontra K  4lDiJcOJ2GLCK6p9q5BgfK   

                                               genres  popularity  \
0                     german cloud rap,german hip hop          72   
1   hip hop,indie pop rap,oakland hip hop,pop rap,rap          83   
2          canadian contemporary r&b,canadian pop,pop          95   
3       frauenrap,german pop,german singer-songwriter          70   
4                  boston hip hop,hip hop,pop rap,rap          78   
5                      german hip hop,hamburg hip hop          70   
6   garage rock,modern rock,permanent wave,rock,sh...          87   
7               adult standards,easy listening,lounge          85   
8             german hip hop,german pop,german reggae          66   
9                 frauenrap,german hip hop,german pop          68   
10  alternative dance,electronica,microhouse,minim...          51   
11                        detroit hip hop,hip hop,rap          94   
12                    christlicher rap,german hip hop          74   

                                      uri  
0   spotify:artist:18ISxWwWjV6rPLoVCXf1dz  
1   spotify:artist:02kJSzxNuaWGqwubyUba0Z  
2   spotify:artist:1Xyo4u8uXC1ZmMpatF05PJ  
3   spotify:artist:65AzRSW0jKSs0WtttEXrOw  
4   spotify:artist:6C1ohJrd5VydigQtaGy5Wa  
5   spotify:artist:5tHiL8SKSaZGMBUPIiSmX4  
6   spotify:artist:7Ln80lUS6He07XvHI8qqHH  
7   spotify:artist:1Mxqyy3pSjf8kZZL4QVxS0  
8   spotify:artist:5ISjkNS17JpCwiFtW80lpV  
9   spotify:artist:4sg4no0TXdsrM1s4SVUwNF  
10  spotify:artist:2jYMYP2SVifgmzNRQJx3SJ  
11  spotify:artist:7dGJo4pcD2V6oG8kP0tJRR  
12  spotify:artist:4lDiJcOJ2GLCK6p9q5BgfK

### Last weeks playlist

In order to avoid encountering the same songs two weeks in a row, which is very likely as short and medium term top tracks won't have changed much, last weeks playlist is read from 'Playlist.csv'. This file was just an empty csv file the first time the code was run. But at the end of the playlist creation process the created playlist is saved in that csv file, so it contains last weeks playlist.

The code that creates the playlist (script_for_duos_playlist.py) does the same every time it is run.

In [15]:
last_week_duo = pd.read_csv(path/"Playlist.csv", index_col = 0)
last_week_duo

track_name                               album  \
0                    Von Party zu Party                      Leben am Limit   
1                            We Love It                 Future Vintage Soul   
2    Take Some Time - Emancipator Remix  Take Some Time (Emancipator Remix)   
3                                  Snub                                Snub   
4             KANN DAS BITTE SO BLEIBEN                               NACHT   
5                               Testo E                             Testo E   
6                       Bande organisée                     Bande organisée   
7                               Fragile                             Fragile   
8                         Beifahrersitz                       Beifahrersitz   
9                               Qa bone                             Qa bone   
10                   Wakka Flocka Flame                  Wakka Flocka Flame   
11                             No Limit              The Beautiful & Damned   
12                             Blackout                         EROS (2018)   
13                              FREUNDE                               NACHT   
14                      Monica Bellucci                         EROS (2018)   
15                              GITARRE                               NACHT   
16                             Oh Junge                         KitschKrieg   
17                     ALL OF MY DREAMS                               NACHT   
18  Lass Das Licht An (feat. Deichkind)                             BAM BAM   
19  Generation YouPorn - Agia Napa Edit    Gefällt euch Faber so besser? EP   
20                      Between Breaths                           Stateless   
21                               Try Me                              Try Me   
22  Good Morning Vietnam - Original Mix                Good Morning Vietnam   
23                               Trauma                              Trauma   
24                               Carmen                         Born To Die   
25                       Nicht Erinnern                      Nicht Erinnern   
26                       Sex & Stardust                      Sex & Stardust   
27                          Nicht jetzt            Maximum (Deluxe Edition)   
28                      Regen im Sommer                               Regen   
29       Lass sie tanzen (Square Dance)                            Euphoria   
30                          Money Money                         Money Money   
31                        Into the Wild              Sapiens Beasts, Vol. 1   
32                               Kleine                               20:14   
33                              Qa bone                           Fast Life   
34                             Jacob&Co                   Came From Nothing   
35               HaifischNikez Allstars              HaifischNikez Allstars   
36             Cash Out - A COLORS SHOW            Cash Out - A COLORS SHOW   
37                                  LFR                                K¡K¡   
38           Erdbeerwoche (feat. Elias)          Erdbeerwoche (feat. Elias)   
39                    Enjoy the Silence                   Enjoy the Silence   
40                         Venice Bitch            Norman Fucking Rockwell!   
41                     Tränen aus Kajal                            Instinkt   

                  track_id            artist  duration  popularity  \
0   4h6UIDvGWTYZvu4BLf2GpO              SXTN    224186          67   
1   3QBTiVakPVVocXZF4H9MQ9         Outasight    189609          28   
2   28M2ugvRSIa4MIKmiiwNao         Wilderado    290428          40   
3   6bc1eT78cvMHU0TuQQaYtU            FLOHIO    134307          23   
4   45HOck8XCgrSlVUQHHOHMz              ELIF    164511          51   
5   0bBhZuQEbfQR4JG9n6BDdP              SSIO    182733          39   
6   205HNJ73cgpC0LAOnuQiWT              Kofs    356346          80   
7   3YYqctc3S1DH1i827bKpAh         Kora (CA)    454008          30   
8

## Creating the Playlist

The creation of the playlist is the main goal of the task and of the project. It requires a couple of steps to 'assemble all building blocks' that make up the DUO.py playlist. THe building blocks are:

1. Common top tracks, that were not in last weeks playlists
2. A sample of each users top tracks that are most similar to the other users top tracks
3. A sample of each users top tracks from an artists both users like (one of their top artists)
4. A sample of the songs saved by users
5. A recommended track (through spotify and additional filtering) for every track that was added to the playlist in step 1.-4.

### Common top tracks

The playlist is initiate by common top tracks, that did not appear in last weeks playlist already. As these songs are both users' favorites they should enjoy them.

In [16]:
dataframe_difference(top_tracks_m,top_tracks_t,which = "both")

C:\Users\ms101\Anaconda3\envs\spotify_app\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


track_name           album                track_id  \
0                          RSVP      Nimmerland  38UlieEW0eto55PNH9Z0cK   
1            Von Party zu Party  Leben am Limit  4h6UIDvGWTYZvu4BLf2GpO   
2  Napauken - Jpattersson Remix           Zehna  2vPP6FNAzYoo2pplHp7Vop   
3                      Bläulich     Treppenhaus  2WRTnY0slmFgWcrmEr8dPj   

       artist  duration  popularity  danceability  energy  key  loudness  \
0         RIN    169293          57         0.787   0.671    9    -4.447   
1        SXTN    224186          67         0.694   0.895    1    -4.349   
2      Shkoon    380616          50         0.793   0.294    9   -13.817   
3  Apache 207    196213          71         0.790   0.704   10    -7.935   

   mode  speechiness  acousticness  instrumentalness  liveness  valence  \
0     1       0.0409        0.0642          0.011200    0.1070    0.548   
1     1       0.2520        0.1770          0.000000    0.0749    0.621   
2     0       0.0556        0.2610          0.900000    0.0867    0.370   
3     0       0.4170        0.0690          0.000658    0.1130    0.212   

     tempo  
0  159.058  
1  120.002  
2  102.029  
3  154.007

In [17]:
common_songs = dataframe_difference(top_tracks_m,top_tracks_t,which = "both")
new_playlist_df = common_songs[~common_songs["track_id"].isin(last_week_duo["track_id"])]
new_playlist_df

C:\Users\ms101\Anaconda3\envs\spotify_app\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


track_name        album                track_id  \
0                          RSVP   Nimmerland  38UlieEW0eto55PNH9Z0cK   
2  Napauken - Jpattersson Remix        Zehna  2vPP6FNAzYoo2pplHp7Vop   
3                      Bläulich  Treppenhaus  2WRTnY0slmFgWcrmEr8dPj   

       artist  duration  popularity  danceability  energy  key  loudness  \
0         RIN    169293          57         0.787   0.671    9    -4.447   
2      Shkoon    380616          50         0.793   0.294    9   -13.817   
3  Apache 207    196213          71         0.790   0.704   10    -7.935   

   mode  speechiness  acousticness  instrumentalness  liveness  valence  \
0     1       0.0409        0.0642          0.011200    0.1070    0.548   
2     0       0.0556        0.2610          0.900000    0.0867    0.370   
3     0       0.4170        0.0690          0.000658    0.1130    0.212   

     tempo  
0  159.058  
2  102.029  
3  154.007

### Users top tracks that are similar to each other.

The next building block consist of songs from both users top tracks that are most similar to on another on the level of audio features.

For this task the top tracks unique to each user are extracted and a similarity matrix is computed. The similarity is based on the audio features (excluding 'key' and 'mode') and computed via cosine similarity (see spotifuncs).
From this matrix the 30 highest similarity scores and corresponding indeces are extracted. The songs corresponding to these indeces are put into a dataframe, any duplicates are dropped and a sample of 10 songs is drawn for the playlist.

Here I am creating a similarity matrix, deleting songs that both dataframes contain first. (common_songs)

In [18]:
unique_top_tracks_m = top_tracks_m[~top_tracks_m["track_id"].isin(common_songs["track_id"])]
unique_top_tracks_m.reset_index(drop = True,inplace = True)
unique_top_tracks_m

track_name  \
0                                              T.H.A.L   
1                                                 Move   
2                                        Rollercoaster   
3                                               Signal   
4                                           You and Me   
..                                                 ...   
102                                         Until Dawn   
103                             INTERNATIONAL GANGSTAS   
104  Points of Authority / 99 Problems / One Step C...   
105                               X Gon' Give It To Ya   
106                                     WAIDMANNS HEIL   

                                       album                track_id  \
0                                    T.H.A.L  73GjtcmkMfHirKnxotxW1Y   
1                                    T.H.A.L  0Ey20aookshI9ZtXc23wxz   
2                                    T.H.A.L  2uWnoJg6wgaSfBoaLPuwfl   
3                                    T.H.A.L  4BG4FBXN3pPhuze6i6lwv8   
4                              Ghetto Cowboy  53QA7j4pHWAtbE3D0Glh7Q   
..                                       ...                     ...   
102                               Until Dawn  1R2rmmUufa1tdiD9j9Qs3x   
103                   INTERNATIONAL GANGSTAS  4IAVIBh9jbYN30DwpHFMD4   
104        Collision Course (Deluxe Version)  65eohvrL4ttjA7EfFkQOhX   
105                          The Best Of DMX  2NeHnSFnwNp1Z5WYgcjJ8L   
106  LIEBE IST FÜR ALLE DA (SPECIAL EDITION)  6ey8jr96hlSORKyeJk9d8d   

         artist  duration  popularity  danceability  energy  key  loudness  \
0    badmómzjay    153975          58         0.919   0.560    9    -7.298   
1    badmómzjay    164360          30         0.837   0.574    3    -5.180   
2    badmómzjay    165833          31         0.815   0.493    2    -9.709   
3    badmómzjay    131055          31         0.834   0.603    3    -6.899   
4      Yelawolf    242001          63         0.598   0.448    1    -8.525   
..          ...       ...         ...           ...     ...  ...       ...   
102      Jaeger    287190          49         0.531   0.643    3    -9.347   
103  Farid Bang    286280          59         0.740   0.622    1    -6.297   
104       JAY-Z    295826          60         0.547   0.951    1    -4.079   
105         DMX    219253           0         0.678   0.857   10    -5.173   
106   Rammstein    212800           0         0.578   0.969    9    -3.894   

     mode  speechiness  acousticness  instrumentalness  liveness  valence  \
0       1       0.3030       0.07710          0.000000    0.0948   0.1940   
1       0       0.2360       0.12800          0.000000    0.1660   0.3450   
2       0       0.0937       0.66200          0.000000    0.1100   0.5440   
3       0       0.1010       0.22200          0.000103    0.0972   0.4100   
4       0       0.0342       0.10900          0.001050    0.1230   0.0727   
..    ...          ...           ...               ...       ...      ...   
102     0       0.0321       0.00221          0.826000    0.0547   0.1150   
103     1       0.2030       0.09650          0.000000    0.0990   0.3930   
104     0       0.1510       0.00272          0.000011    0.0862   0.5100   
105     0       0.2160       0.03250          0.000000    0.0788   0.6220   
106     1       0.0317       0.00113          0.023400    0.0579   0.6840   

       tempo  
0    139.978  
1     95.987  
2    142.112  
3    104.087  
4    140.970  
..       ...  
102  150.087  
103  148.040  
104   95.031  
105   94.950  
106  104.999  

[107 rows x 17 columns]

In [19]:
unique_top_tracks_t = top_tracks_t[~top_tracks_t["track_id"].isin(common_songs["track_id"])]
unique_top_tracks_t.reset_index(drop = True,inplace = True)
unique_top_tracks_t

track_name  \
0                                         Bati   
1                               Intro (Megbia)   
2                                      Sai Bai   
3          Last Night in Sant Celoni - 12" Mix   
4                                      Anemogn   
..                                         ...   
100                       The Shape - Acoustic   
101                             Painting Greys   
102  Comets (feat. Natalia Doco) - HUGEL Remix   
103                               Worlds Apart   
104                             Double in Love   

                                           album                track_id  \
0                                Chelina, Vol. 1  03xV05Oll19y59x8GDkWVL   
1                                Chelina, Vol. 1  0aLjvBLCCpNNoBwGqNY6Gn   
2                                Chelina, Vol. 1  3vAJlGxRNNYYMBJdnhxynu   
3    Last Night in Sant Celoni (feat. Jaz James)  6zAqmi9YkgnMxbqLcf7hDv   
4                                Chelina, Vol. 1  1BXDo9qPfakhiJfiLZYJ1w   
..                                           ...                     ...   
100                         The Shape (Acoustic)  0lReYwQUfeLD2O4ZWlj5lr   
101                                     Prologue  2Eh0J9SM2SzKnw5spcHJ1B   
102    Comets (feat. Natalia Doco) [The Remixes]  1SadOe3BGZW9MlW2OgQJyd   
103                                 Worlds Apart  5L4YMktQZVQcVZnCHFNymD   
104                       Sticker on My Suitcase  0YfEgbU4mcFQJNHVP3sKsd   

             artist  duration  popularity  danceability  energy  key  \
0           Chelina    260713          12         0.575  0.3610    2   
1           Chelina     53034           6         0.554  0.0872    9   
2           Chelina    201875          23         0.695  0.6770    1   
3           Payfone    427150          48         0.823  0.4360    9   
4           Chelina    193714           7         0.614  0.3060    2   
..              ...       ...         ...           ...     ...  ...   
100  Nico de Andrea    166946          12         0.433  0.5090    8   
101      Emmit Fenn    227368          53         0.870  0.2220    4   
102   Freddy Verano    266218          27         0.844  0.5620   11   
103             Ash    263367          17         0.707  0.5570    5   
104     Alle Farben    159000          32         0.821  0.6740    9   

     loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
0      -8.074     0       0.0291        0.5840          0.010300    0.1170   
1     -17.823     0       0.1510        0.9090          0.000000    0.1160   
2      -4.925     0       0.0656        0.0364          0.006430    0.1810   
3     -10.439     0       0.1050        0.6820          0.051900    0.3430   
4      -7.428     1       0.0350        0.8570          0.000002    0.1220   
..        ...   ...          ...           ...               ...       ...   
100    -7.544     0       0.0320        0.5750          0.000000    0.1070   
101   -14.323     0       0.0766        0.4570          0.036400    0.0909   
102    -6.323     0       0.0932        0.2580          0.031600    0.6220   
103    -8.848     0       0.0356        0.1850          0.901000    0.0999   
104    -6.728     0       0.0572        0.1180          0.001450    0.0351   

     valence    tempo  
0      0.430  148.154  
1      0.328   68.149  
2      0.394  176.174  
3      0.452  105.096  
4      0.169   69.939  
..       ...      ...  
100    0.314  102.644  
101    0.199  132.988  
102    0.385  118.982  
103    0.124  113.023  
104    0.689  120.000  

[105 rows x 17 columns]

In [20]:
similarity_top_songs = create_similarity_score(unique_top_tracks_m,unique_top_tracks_t)

In [21]:
similarity_top_songs

array([[0.76758631, 0.50963858, 0.83814992, ..., 0.82878224, 0.7118232 ,
        0.85323572],
       [0.76986956, 0.5656348 , 0.80342259, ..., 0.86082282, 0.70061781,
        0.88419206],
       [0.9264914 , 0.8100283 , 0.76695033, ..., 0.80575341, 0.65282266,
        0.85061301],
       ...,
       [0.72853323, 0.36694645, 0.84462263, ..., 0.79169662, 0.69107247,
        0.92391296],
       [0.75786834, 0.4580677 , 0.84583769, ..., 0.81988046, 0.69204618,
        0.95789938],
       [0.75286728, 0.35877836, 0.85198439, ..., 0.78001165, 0.69013424,
        0.94992055]])

Creating a list of tuples containing the indeces for both songs as well as the similarity score.

In [22]:
max_n_scores = [(i,np.argmax(x),x[np.argmax(x)]) for i,x in enumerate(similarity_top_songs)]
max_n_scores

[(0, 5, 0.9419454823140679),
 (1, 75, 0.9278199095090397),
 (2, 56, 0.977899055967248),
 (3, 17, 0.9509538090620334),
 (4, 35, 0.9145349845034312),
 (5, 91, 0.9824336825533914),
 (6, 61, 0.8219181556472046),
 (7, 3, 0.950700535532872),
 (8, 45, 0.9755812907458108),
 (9, 60, 0.9361726243700328),
 (10, 17, 0.979407756162956),
 (11, 8, 0.9363536976479466),
 (12, 76, 0.9778383926517018),
 (13, 90, 0.977080665933461),
 (14, 45, 0.9420800032333909),
 (15, 17, 0.9967309242459156),
 (16, 66, 0.9610589046819443),
 (17, 6, 0.9844914310152725),
 (18, 17, 0.9757130646680877),
 (19, 75, 0.973887748190232),
 (20, 75, 0.9396202108425412),
 (21, 98, 0.9882028630651086),
 (22, 17, 0.9701554122027172),
 (23, 21, 0.9460241499463621),
 (24, 7, 0.963035438385104),
 (25, 20, 0.9770524509266321),
 (26, 58, 0.9801418832005209),
 (27, 75, 0.9869318137395139),
 (28, 103, 0.9958680223280652),
 (29, 75, 0.986822958720218),
 (30, 98, 0.9706459428460529),
 (31, 58, 0.9354849072064834),
 (32, 64, 0.9732298475604584)

In [23]:
from operator import itemgetter
from heapq import nlargest
nlargest(30,max_n_scores,key=itemgetter(2))

[(15, 17, 0.9967309242459156),
 (28, 103, 0.9958680223280652),
 (38, 75, 0.990010468350755),
 (57, 80, 0.9889265279367138),
 (101, 16, 0.9888603135251002),
 (21, 98, 0.9882028630651086),
 (105, 75, 0.9872399036503025),
 (50, 79, 0.9871812260261),
 (86, 104, 0.9871627741414891),
 (27, 75, 0.9869318137395139),
 (29, 75, 0.986822958720218),
 (72, 93, 0.9849107732495253),
 (17, 6, 0.9844914310152725),
 (74, 79, 0.9839898075777093),
 (5, 91, 0.9824336825533914),
 (62, 73, 0.9821532025817385),
 (49, 68, 0.9819266545196924),
 (71, 76, 0.9814452587822746),
 (93, 76, 0.9810067008687731),
 (47, 63, 0.9806737828245374),
 (26, 58, 0.9801418832005209),
 (10, 17, 0.979407756162956),
 (102, 103, 0.9793002550763088),
 (33, 80, 0.9786279448999823),
 (2, 56, 0.977899055967248),
 (55, 9, 0.9778466897640011),
 (12, 76, 0.9778383926517018),
 (60, 17, 0.9776281556554647),
 (46, 103, 0.9772706099902545),
 (13, 90, 0.977080665933461)]

Extracting the track pairs with the 30 highest similarity scores for each user.

In [24]:
idx_simtracks_m = [i[0] for i in  nlargest(30,max_n_scores,key=itemgetter(2))]
idx_simtracks_t = [i[1] for i in  nlargest(30,max_n_scores,key=itemgetter(2))]

In [25]:
idx_simtracks_m

[15,
 28,
 38,
 57,
 101,
 21,
 105,
 50,
 86,
 27,
 29,
 72,
 17,
 74,
 5,
 62,
 49,
 71,
 93,
 47,
 26,
 10,
 102,
 33,
 2,
 55,
 12,
 60,
 46,
 13]

In [26]:
idx_simtracks_t

[17,
 103,
 75,
 80,
 16,
 98,
 75,
 79,
 104,
 75,
 75,
 93,
 6,
 79,
 91,
 73,
 68,
 76,
 76,
 63,
 58,
 17,
 103,
 80,
 56,
 9,
 76,
 17,
 103,
 90]

In [27]:
sim_top_tracks_m = unique_top_tracks_m.loc[idx_simtracks_m]
sim_top_tracks_t = unique_top_tracks_t.loc[idx_simtracks_t]

Creating the dataframe with the most similar top tracks.

In [28]:
similar_top_tracks = pd.concat([sim_top_tracks_m,sim_top_tracks_t])
similar_top_tracks.drop_duplicates(inplace = True)
similar_top_tracks = similar_top_tracks[~similar_top_tracks["track_id"].isin(last_week_duo["track_id"])]
similar_top_tracks.reset_index(drop = True,inplace = True)
similar_top_tracks

track_name  \
0                               Vagabundo   
1                           Another Place   
2                               Hurricane   
3                   Burn (feat. Big Sean)   
4                    Who - Single Version   
5                    X Gon' Give It To Ya   
6                      FEIND (feat. Azad)   
7                               Ausziehen   
8                                   Scary   
9                  AUGEN ZU (feat. Samra)   
10  For an Endless Night - Jel Ford Remix   
11                     Lost In The Static   
12                Movin' Bass - GTA Remix   
13                Movin' Bass - GTA Remix   
14                                    069   
15                               Irrbloss   
16                                  sugar   
17                             Until Dawn   
18              Struggle Made Me Stronger   
19                          Rollercoaster   
20              Berlin lebt wie nie zuvor   
21                           Deine Mutter   
22                        ABER WO BIST DU   
23                              Southerly   
24                                Pandora   
25                               Beyikrta   
26                           Worlds Apart   
27                 Если ты меня не любишь   
28                                    505   
29                                  Gonay   
30                                   Home   
31                             Grey Goose   
32                         Double in Love   
33                                  Creep   
34                     IC3 (feat. Skepta)   
35                                  Drugs   
36                 Huldra - Other Version   
37                      Keiner hat gesagt   
38                                  Wanaw   
39                               Kalimera   

                                     album                track_id  \
0                              EROS (2018)  4MRitCc3SHpBsqbNbHysm8   
1                            Another Place  1cyRXghdofqkqavGpQzHZt   
2                                   TRAUMA  2GFwwTIVLjnOrtP7m9luHC   
3   Dreams and Nightmares (Deluxe Edition)  1tv6IvWoOilhj0XbBoVVMo   
4                                      Who  0CjBORMsmiQNe3vPDcNIvk   
5                          The Best Of DMX  2NeHnSFnwNp1Z5WYgcjJ8L   
6                                    NACHT  4Upt2Q8OcUd85uo5x9QqWK   
7                           Leben am Limit  1U2fPOdYmSz44FLUaboG9M   
8                                    Scary  745Dazbwplj1SDZ8SPKHV5   
9                                    NACHT  6ikiAn9th3TUfS3bYS7gDX   
10   For an Endless Night (Jel Ford Remix)  39KN3wgqTrGFxvWRyMW5zL   
11                                Dig Deep  5QLH7zAdcAJLgR25gtvtoK   
12                 Movin' Bass (GTA Remix)  7yzwszgJUDrllWnG9cQkxQ   
13                 Movin' Bass (GTA Remix)  2BnVVl5NCV2o7XlZo1CLEm   
14                              Unzensiert  1B2s3CpX2PV0YN7DkLSQ4b   
15                                Irrbloss  1Brs2yGNA859g7wx3xojOM   
16                            next chapter  5IjIbGO7lih9CVDBFLCtTT   
17                              Until Dawn  1R2rmmUufa1tdiD9j9Qs3x   
18                 Headphones on World Off  3vMtLNMANvkPuDthcLuQzJ   
19                                 T.H.A.L  2uWnoJg6wgaSfBoaLPuwfl   
20                           Berlin lebt 2  2pfio0uHpT4USTPBNeTIWo   
21                            Deine Mutter  13xrrko52JwoBc0AS17ezX   
22                                   NACHT  2dGtLJysTSI9cbQ6TulL8V   
23                               Southerly  2GwpIphHopWzZgHI7m4rnZ   
24                                 Pandora  2kscYCgdy8FP8D92hgsVIy   
25                         Chelina, Vol. 1  6Kd77EjLuU4b8NZyOEkyTu   
26                            Worlds Apart  5L4YMktQZVQcVZnCHFNymD   
27                          Косатка в небе  04tIaPPuw8ecPBfWgOZEzn   
28               Favourite Worst Nightmare  0BxE4FqsDD1Ot4YuBXwAPp   
29                         Chelina, Vol. 1  64CTlQomSyNlopchI1U7lg   
30             

In [29]:
similar_top_tracks.sample(10)

track_name  \
33                                  Creep   
11                     Lost In The Static   
22                        ABER WO BIST DU   
10  For an Endless Night - Jel Ford Remix   
20              Berlin lebt wie nie zuvor   
9                  AUGEN ZU (feat. Samra)   
3                   Burn (feat. Big Sean)   
32                         Double in Love   
23                              Southerly   
34                     IC3 (feat. Skepta)   

                                     album                track_id  \
33                             Pablo Honey  6b2oQwSGFkzsMtQruIWm2p   
11                                Dig Deep  5QLH7zAdcAJLgR25gtvtoK   
22                                   NACHT  2dGtLJysTSI9cbQ6TulL8V   
10   For an Endless Night (Jel Ford Remix)  39KN3wgqTrGFxvWRyMW5zL   
20                           Berlin lebt 2  2pfio0uHpT4USTPBNeTIWo   
9                                    NACHT  6ikiAn9th3TUfS3bYS7gDX   
3   Dreams and Nightmares (Deluxe Edition)  1tv6IvWoOilhj0XbBoVVMo   
32                  Sticker on My Suitcase  0YfEgbU4mcFQJNHVP3sKsd   
23                               Southerly  2GwpIphHopWzZgHI7m4rnZ   
34                      IC3 (feat. Skepta)  0BMszcKwCrUDHZX3CEEj6L   

              artist  duration  popularity  danceability  energy  key  \
33         Radiohead    238640          69         0.515   0.430    7   
11  After The Burial    273194          60         0.602   0.955    6   
22              ELIF    185274          56         0.780   0.592    9   
10  Alan Fitzpatrick    449768          39         0.751   0.649    8   
20       Capital Bra    151294          59         0.654   0.782    1   
9               ELIF    174248          71         0.685   0.693    5   
3          Meek Mill    216666          55         0.670   0.895    2   
32       Alle Farben    159000          32         0.821   0.674    9   
23           Tom Day    284709          33         0.503   0.778    0   
34            Ghetts    231106          58         0.663   0.645    8   

    loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
33    -9.935     1       0.0369      0.010200          0.000141    0.1290   
11    -4.888     1       0.1540      0.000989          0.013300    0.2070   
22    -6.020     0       0.0631      0.230000          0.000000    0.1840   
10    -6.885     0       0.0579      0.015100          0.761000    0.0976   
20    -2.880     1       0.2440      0.198000          0.000000    0.3470   
9     -6.741     0       0.2170      0.125000          0.000002    0.2750   
3     -1.438     1       0.3490      0.327000          0.000000    0.2690   
32    -6.728     0       0.0572      0.118000          0.001450    0.0351   
23   -10.125     1       0.0443      0.250000          0.855000    0.1290   
34    -7.977     1       0.3980      0.443000          0.000000    0.1660   

    valence    tempo  
33    0.104   91.841  
11    0.193  126.009  
22    0.705  148.057  
10    0.179  125.996  
20    0.635  179.701  
9     0.493  167.957  
3     0.622  111.389  
32    0.689  120.000  
23    0.300  138.001  
34    0.394  128.023

In [30]:
new_playlist_df = new_playlist_df.append(similar_top_tracks.sample(10))
new_playlist_df

track_name                    album  \
0                           RSVP               Nimmerland   
2   Napauken - Jpattersson Remix                    Zehna   
3                       Bläulich              Treppenhaus   
22               ABER WO BIST DU                    NACHT   
17                    Until Dawn               Until Dawn   
19                 Rollercoaster                  T.H.A.L   
18     Struggle Made Me Stronger  Headphones on World Off   
25                      Beyikrta          Chelina, Vol. 1   
5           X Gon' Give It To Ya          The Best Of DMX   
20     Berlin lebt wie nie zuvor            Berlin lebt 2   
33                         Creep              Pablo Honey   
34            IC3 (feat. Skepta)       IC3 (feat. Skepta)   
6             FEIND (feat. Azad)                    NACHT   

                  track_id               artist  duration  popularity  \
0   38UlieEW0eto55PNH9Z0cK                  RIN    169293          57   
2   2vPP6FNAzYoo2pplHp7Vop               Shkoon    380616          50   
3   2WRTnY0slmFgWcrmEr8dPj           Apache 207    196213          71   
22  2dGtLJysTSI9cbQ6TulL8V                 ELIF    185274          56   
17  1R2rmmUufa1tdiD9j9Qs3x               Jaeger    287190          49   
19  2uWnoJg6wgaSfBoaLPuwfl           badmómzjay    165833          31   
18  3vMtLNMANvkPuDthcLuQzJ  Fearless Motivation    213000          48   
25  6Kd77EjLuU4b8NZyOEkyTu              Chelina    216410           7   
5   2NeHnSFnwNp1Z5WYgcjJ8L                  DMX    219253           0   
20  2pfio0uHpT4USTPBNeTIWo          Capital Bra    151294          59   
33  6b2oQwSGFkzsMtQruIWm2p            Radiohead    238640          69   
34  0BMszcKwCrUDHZX3CEEj6L               Ghetts    231106          58   
6   4Upt2Q8OcUd85uo5x9QqWK                 ELIF    169299          53   

    danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.787   0.671    9    -4.447     1       0.0409      0.064200   
2          0.793   0.294    9   -13.817     0       0.0556      0.261000   
3          0.790   0.704   10    -7.935     0       0.4170      0.069000   
22         0.780   0.592    9    -6.020     0       0.0631      0.230000   
17         0.531   0.643    3    -9.347     0       0.0321      0.002210   
19         0.815   0.493    2    -9.709     0       0.0937      0.662000   
18         0.358   0.886    9    -4.480     0       0.2500      0.000645   
25         0.768   0.467   10    -8.708     0       0.0491      0.321000   
5          0.678   0.857   10    -5.173     0       0.2160      0.032500   
20         0.654   0.782    1    -2.880     1       0.2440      0.198000   
33         0.515   0.430    7    -9.935     1       0.0369      0.010200   
34         0.663   0.645    8    -7.977     1       0.3980      0.443000   
6          0.781   0.770    1    -5.367     1       0.0482      0.072300   

    instrumentalness  liveness  valence    tempo  
0           0.011200    0.1070    0.548  159.058  
2           0.900000    0.0867    0.370  102.029  
3           0.000658    0.1130    0.212  154.007  
22          0.000000    0.1840    0.705  148.057  
17          0.826000    0.0547    0.115  150.087  
19          0.000000    0.1100    0.544  142.112  
18          0.000000    0.1130    0.238  173.965  
25          0.000000    0.1570    0.616   90.045  
5           0.000000    0.0788    0.622   94.950  
20          0.000000    0.3470    0.635  179.701  
33          0.000141    0.1290    0.104   91.841  
34          0.000000    0.1660    0.394  128.023  
6           0.000008    0.1030    0.595  148.005

### Sampling from each users top tracks which are from an artists both users like (one of their top artists)

The next step filters the top tracks based on common artists. Every track by an artist that is a top artist of **both** users is considered for this approach.



In [31]:
filtered_top_m = top_tracks_m[top_tracks_m["artist"].isin(common_artists["name"]) 
                              & ~top_tracks_m["track_id"].isin(last_week_duo["track_id"])]
filtered_top_m

track_name                album  \
5                                              RSVP           Nimmerland   
8                         Aretha Franklin Freestyle          EROS (2018)   
16                                        Vagabundo          EROS (2018)   
22                             Who - Single Version                  Who   
52                                      ALLES HELAL                NACHT   
58                                            FEUER                NACHT   
62                                          SCHWARZ                NACHT   
63                                           ALASKA                NACHT   
64                                  ABER WO BIST DU                NACHT   
65                                             ADHD                 ADHD   
70                                  EIN LETZTES MAL                NACHT   
71                                      Lambo Lambo          KitschKrieg   
77                                          NUR MIR                NACHT   
78                           AUGEN ZU (feat. Samra)                NACHT   
79                                      Fall Slowly          Fall Slowly   
85                            DU TUST MIR NICHT GUT                NACHT   
90                               FEIND (feat. Azad)                NACHT   
92                                       OHNE LIEBE                NACHT   
93   Fall Slowly (feat. Ashanti) - Extended Version            Evolution   
94                                        Evolution            Evolution   
104                          I Mean It (feat. Remo)  These Things Happen   

                   track_id        artist  duration  popularity  danceability  \
5    38UlieEW0eto55PNH9Z0cK           RIN    169293          57         0.787   
8    5hKM3DaySvongkyGdvhacX           RIN    184586          55         0.790   
16   4MRitCc3SHpBsqbNbHysm8           RIN    125483          43         0.878   
22   0CjBORMsmiQNe3vPDcNIvk  Modeselektor    207988          44         0.760   
52   2KAbQ3PsETrr86R39pru7k          ELIF    175062          62         0.727   
58   0Se4w42WIJiTgld4SYbv8S          ELIF    203386          47         0.769   
62   7vswtCdKBzC5XN9ojwh8u0          ELIF    135184          47         0.742   
63   2YnYp5f38UP6fvf7q2FnPm          ELIF    217766          49         0.743   
64   2dGtLJysTSI9cbQ6TulL8V          ELIF    185274          56         0.780   
65   4X4v3KtkUXwXvDBw5KS9cp  Joyner Lucas    205872          72         0.563   
70   04EJyZSlhPPfFOo1NRn2vl          ELIF    144644          52         0.713   
71   7oqvRZNv4dUV8CgQWtIAMe   KitschKrieg    214991          59         0.876   
77   1jzoAcvAE7vykG1sTI2mvt          ELIF    150136          48         0.737   
78   6ikiAn9th3TUfS3bYS7gDX          ELIF    174248          71         0.685   
79   01WOwxkxOw2FqNIHkraxcN  Joyner Lucas    208095          63         0.576   
85   2pmqIWXxhdvjJfRiwySIZl          ELIF    207279          54         0.860   
90   4Upt2Q8OcUd85uo5x9QqWK          ELIF    169299          53         0.781   
92   6lX6V9FUuix1ZQBX4Vngwc          ELIF    194975          48         0.714   
93   7wK4pOTZKVaAJ00rziu901  Joyner Lucas    292056          57         0.542   
94   2VopDw2GlF3uwD1kihHmTT  Joyner Lucas    153250          61         0.687   
104  6jmTHeoWvBaSrwWttr8Xvu        G-Eazy    236480          75         0.712   

     energy  key  loudness  mode  speechiness  acousticness  instrumentalness  \
5     0.671    9    -4.447     1       0.0409       0.06420          0.011200   
8     0.592    3    -7.226     0       0.0811       0.49800          0.325000   
16    0.714    6    -8.036     0       0.1290       0.37500          0.000009   
22    0.910    7    -8.472     1       0.0484       0.04420          0.006620   
52    0.600    4    -6.186     0       0.0376       0.16400          0.000017   
58    0.604    8    -6.769     1       0.0455       0.03900          0.000277   
62    0.700   10  

In [32]:
filtered_top_t = top_tracks_t[top_tracks_t["artist"].isin(common_artists["name"])
                             & ~top_tracks_t["track_id"].isin(last_week_duo["track_id"])]
filtered_top_t

track_name                      album                track_id  \
13           RSVP                 Nimmerland  38UlieEW0eto55PNH9Z0cK   
40         M.I.A.                 Nimmerland  06pbKFx2Iut0MRs1XjV9Mc   
50         Wealth                   Who Else  5aOlYhQsp75cgPov4yjWIe   
83            505  Favourite Worst Nightmare  0BxE4FqsDD1Ot4YuBXwAPp   
98  Lass Sie Gehn                    BAM BAM  4RYKr1R3tXrITqY1zWiTNi   

            artist  duration  popularity  danceability  energy  key  loudness  \
13             RIN    169293          57         0.787   0.671    9    -4.447   
40             RIN    190640          50         0.663   0.691    4    -3.903   
50    Modeselektor    247218          43         0.765   0.452    1   -12.346   
83  Arctic Monkeys    253586          82         0.526   0.866    0    -5.822   
98           Seeed    192853          55         0.711   0.706    7    -4.543   

    mode  speechiness  acousticness  instrumentalness  liveness  valence  \
13     1       0.0409       0.06420          0.011200    0.1070    0.548   
40     1       0.0586       0.05560          0.000005    0.1040    0.190   
50     1       0.0856       0.02040          0.000057    0.0727    0.331   
83     1       0.0568       0.00287          0.000078    0.0945    0.248   
98     1       0.2890       0.07760          0.000000    0.1270    0.695   

      tempo  
13  159.058  
40  161.866  
50  137.972  
83  140.266  
98  138.810

#### Potential issues

This filtering approach often leads to lists of songs that only contain very few artists but a couple of songs by that artist (it's also due to spotify really noticing when you can't stop listening to an album..)
To not have too many songs by the same artist I will sample from the above dataframes.I am assigning weights to the rows depending on how often an artist occurs.

This approach worked reasonably well however still has some flaws (which might be partially driven by my girlfriends and my individual listening behavior)

In [33]:
from collections import Counter

Counter(filtered_top_m["artist"]) ,  Counter(filtered_top_t["artist"])

(Counter({'RIN': 3,
          'Modeselektor': 1,
          'ELIF': 11,
          'Joyner Lucas': 4,
          'KitschKrieg': 1,
          'G-Eazy': 1}),
 Counter({'RIN': 2, 'Modeselektor': 1, 'Arctic Monkeys': 1, 'Seeed': 1}))

In [34]:
weights_m = [1/len(filtered_top_m)/7 if Counter(filtered_top_m["artist"])[x] > 2 else 1/len(filtered_top_m) for x in filtered_top_m["artist"]] 

In [35]:
weights_t = [1/len(filtered_top_t)/7 if Counter(filtered_top_t["artist"])[x] > 2 else 1/len(filtered_top_m) for x in filtered_top_t["artist"]] 

I tried the sampling with and without weights 10 times. With weights artists that occur very often in the filtered dataframe are not too overrepresented just as planned. Without weights the sample sometimes contained just one or two artists which is not desired.

In [36]:
sample_n = (25-len(new_playlist_df))//2
if sample_n > 3: sample_n = 3
sample_n

3

In [37]:
filtered_top_m.sample(sample_n,weights = weights_m)

track_name                album                track_id  \
104  I Mean It (feat. Remo)  These Things Happen  6jmTHeoWvBaSrwWttr8Xvu   
70          EIN LETZTES MAL                NACHT  04EJyZSlhPPfFOo1NRn2vl   
71              Lambo Lambo          KitschKrieg  7oqvRZNv4dUV8CgQWtIAMe   

          artist  duration  popularity  danceability  energy  key  loudness  \
104       G-Eazy    236480          75         0.712   0.562   10    -6.008   
70          ELIF    144644          52         0.713   0.549    1    -6.168   
71   KitschKrieg    214991          59         0.876   0.400    5    -9.748   

     mode  speechiness  acousticness  instrumentalness  liveness  valence  \
104     1       0.1290        0.1250          0.000000     0.136    0.142   
70      1       0.0645        0.0461          0.000041     0.108    0.393   
71      0       0.1290        0.4940          0.000009     0.100    0.484   

       tempo  
104  140.000  
70    89.005  
71   144.938

In [38]:
filtered_top_t.sample(sample_n, weights= weights_t)

track_name                      album                track_id  \
13       RSVP                 Nimmerland  38UlieEW0eto55PNH9Z0cK   
83        505  Favourite Worst Nightmare  0BxE4FqsDD1Ot4YuBXwAPp   
40     M.I.A.                 Nimmerland  06pbKFx2Iut0MRs1XjV9Mc   

            artist  duration  popularity  danceability  energy  key  loudness  \
13             RIN    169293          57         0.787   0.671    9    -4.447   
83  Arctic Monkeys    253586          82         0.526   0.866    0    -5.822   
40             RIN    190640          50         0.663   0.691    4    -3.903   

    mode  speechiness  acousticness  instrumentalness  liveness  valence  \
13     1       0.0409       0.06420          0.011200    0.1070    0.548   
83     1       0.0568       0.00287          0.000078    0.0945    0.248   
40     1       0.0586       0.05560          0.000005    0.1040    0.190   

      tempo  
13  159.058  
83  140.266  
40  161.866

In [39]:
new_playlist_df = new_playlist_df.append(filtered_top_m.sample(sample_n,weights = weights_m))
new_playlist_df = new_playlist_df.append(filtered_top_t.sample(sample_n,weights = weights_t))

In [40]:
new_playlist_df = new_playlist_df.drop_duplicates().reset_index(drop=True)
new_playlist_df

track_name                      album  \
0                           RSVP                 Nimmerland   
1   Napauken - Jpattersson Remix                      Zehna   
2                       Bläulich                Treppenhaus   
3                ABER WO BIST DU                      NACHT   
4                     Until Dawn                 Until Dawn   
5                  Rollercoaster                    T.H.A.L   
6      Struggle Made Me Stronger    Headphones on World Off   
7                       Beyikrta            Chelina, Vol. 1   
8           X Gon' Give It To Ya            The Best Of DMX   
9      Berlin lebt wie nie zuvor              Berlin lebt 2   
10                         Creep                Pablo Honey   
11            IC3 (feat. Skepta)         IC3 (feat. Skepta)   
12            FEIND (feat. Azad)                      NACHT   
13                   Lambo Lambo                KitschKrieg   
14          Who - Single Version                        Who   
15        AUGEN ZU (feat. Samra)                      NACHT   
16                           505  Favourite Worst Nightmare   
17                        Wealth                   Who Else   
18                        M.I.A.                 Nimmerland   

                  track_id               artist  duration  popularity  \
0   38UlieEW0eto55PNH9Z0cK                  RIN    169293          57   
1   2vPP6FNAzYoo2pplHp7Vop               Shkoon    380616          50   
2   2WRTnY0slmFgWcrmEr8dPj           Apache 207    196213          71   
3   2dGtLJysTSI9cbQ6TulL8V                 ELIF    185274          56   
4   1R2rmmUufa1tdiD9j9Qs3x               Jaeger    287190          49   
5   2uWnoJg6wgaSfBoaLPuwfl           badmómzjay    165833          31   
6   3vMtLNMANvkPuDthcLuQzJ  Fearless Motivation    213000          48   
7   6Kd77EjLuU4b8NZyOEkyTu              Chelina    216410           7   
8   2NeHnSFnwNp1Z5WYgcjJ8L                  DMX    219253           0   
9   2pfio0uHpT4USTPBNeTIWo          Capital Bra    151294          59   
10  6b2oQwSGFkzsMtQruIWm2p            Radiohead    238640          69   
11  0BMszcKwCrUDHZX3CEEj6L               Ghetts    231106          58   
12  4Upt2Q8OcUd85uo5x9QqWK                 ELIF    169299          53   
13  7oqvRZNv4dUV8CgQWtIAMe          KitschKrieg    214991          59   
14  0CjBORMsmiQNe3vPDcNIvk         Modeselektor    207988          44   
15  6ikiAn9th3TUfS3bYS7gDX                 ELIF    174248          71   
16  0BxE4FqsDD1Ot4YuBXwAPp       Arctic Monkeys    253586          82   
17  5aOlYhQsp75cgPov4yjWIe         Modeselektor    247218          43   
18  06pbKFx2Iut0MRs1XjV9Mc                  RIN    190640          50   

    danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.787   0.671    9    -4.447     1       0.0409      0.064200   
1          0.793   0.294    9   -13.817     0       0.0556      0.261000   
2          0.790   0.704   10    -7.935     0       0.4170      0.069000   
3          0.780   0.592    9    -6.020     0       0.0631      0.230000   
4          0.531   0.643    3    -9.347     0       0.0321      0.002210   
5          0.815   0.493    2    -9.709     0       0.0937      0.662000   
6          0.358   0.886    9    -4.480     0       0.2500      0.000645   
7          0.768   0.467   10    -8.708     0       0.0491      0.321000   
8          0.678   0.857   10    -5.173     0       0.2160      0.032500   
9          0.654   0.782    1    -2.880     1       0.2440      0.198000   
10         0.515   0.430    7    -9.935     1       0.0369      0.010200   
11         0.663   0.645    8    -7.977     1       0.3980      0.443000   
12         0.781   0.770    1    -5.367     1       0.0482      0.072300   
13         0.876   0.400    5    -9.748     0       0.1290      0.494000   
14         0.760   0.910    7    -8.472     1       0.0484      0.044200   
15         0.685   0.693    5    -6.741     0       0.2170      0.125000   
16      

### Sampling from saved tracks

I am aiming for around 25 known tracks (and 25 new ones through recommendations). To achieve this and to somehow account for the somewhat random nature of the previous steps I am filling the playlist with sampled saved tracks.

In [41]:
#sample the remaining 25-len(new_playlist_df) from saved_tracks
#first get audio_features
saved_tracks_m = append_audio_features(saved_tracks_m, sp_m)
saved_tracks_t = append_audio_features(saved_tracks_t,sp_t)

In [42]:
#filter again so artists are not already in new_playlist_df
filtered_saved_m = saved_tracks_m[~saved_tracks_m["artist"].isin(new_playlist_df["artist"])]
filtered_saved_m

track_name  \
0                                         Aufbruch   
1                                           Try Me   
2                                     From America   
3                                   I Fell In Love   
4                                    The Beginning   
5                                  Between Breaths   
6                                    Berlin Nights   
7   Fall Slowly (feat. Ashanti) - Extended Version   
8                                        Evolution   
9                                        Zim Zimma   
10        Money In The Grave (Drake ft. Rick Ross)   
11                                   Beifahrersitz   
12                                          Twingo   
13                                            ADHD   
14                                     Fall Slowly   
15                                  When I'm Small   
16                  Risky Business - Mathame Remix   
17                                            Aura   
20                                     Ghostkeeper   
21                                         Falling   
22                     Five Minutes - Zimmer Remix   
23                                    Five Minutes   
24          The Man With the Red Face - Video Edit   
25              Man With The Red Face - Radio Edit   
26                                  Slap the Ghost   
27                               You.Won't.Do.Shit   
28                                      Full Metal   
29                              Lost In The Static   
30                 Bounce Back (feat. Jay Anthony)   
31                                          So Far   
32                          Idol (feat. Tech N9ne)   
33                                         Phoenix   
34                                            YAYA   
35                                      Light Rain   
36                                        Over Now   
37                                Grave - Stripped   
38                                           ERROR   
39                                          Killer   
40                                         BURNING   
41                                           Alarm   
42                              Stargazing - Remix   
43                                          Omertà   
44             Generation YouPorn - Agia Napa Edit   
45                                    C'est La Vie   

                                              album                track_id  \
0                                          Aufbruch  4aH7Y1gpgLPlstw81dUpDE   
1                                            Try Me  72XJZqZHYjnUVGOXZJX8dA   
2                                     Play Me Again  5BMg8D9Wl4yvPqzTq7rWRC   
3                                      Autumn Bells  0zrHx4EhLePXUa8KhGpS3E   
4                                     The Beginning  4njhWDhTAjhReWtYkiMH9t   
5                                         Stateless  4DJGTc1OsgqFsUGI6W8Mtx   
6                                     Berlin Nights  6gG1R1bFdJeNc2ERAwXxCb   
7                                         Evolution  7wK4pOTZKVaAJ00rziu901   
8                                         Evolution  2VopDw2GlF3uwD1kihHmTT   
9                                         Evolution  17nPeSliosCi427f0lUb75   
10                       The Best In The World Pack  5ry2OE6R2zPQFDO85XkgRb   
11                                    Beifahrersitz  01qOl2pM8emx1sxdBQc05g   
12                                              POP  6vFEvkXOnOTUMgPHxogIRK   
13                                             ADHD  4X4v3KtkUXwXvDBw5KS9cp   
14                                      Fall Slowly  01WOwxkxOw2FqNIHkraxcN   
15                                    Eyelid Movies  3498wF96LsgVgMkGmJzJOC   
16                   Risky Business (Mathame Remix)  5R4hprpCcdgKz1DsPoh9p2   
17                         Aura (Exclusive Version)  6sDb7wNlVXQGqnhEHiNt8B   
20                                      Ghostkeeper  13EMVBU1hcy5yJdX1Xhf9a   
21                           

In [43]:
filtered_saved_t = saved_tracks_t[~saved_tracks_t["artist"].isin(new_playlist_df["artist"])]
filtered_saved_t

track_name  \
0                                     Вою на луну   
2                                           Light   
3                                            Oboe   
4             Good Morning Vietnam - Original Mix   
5                                     Jazz Got Me   
6                                          Trauma   
7                                      By the Sea   
8           From the Beginning - Extended Version   
9                             Hypnotised - EP Mix   
10                                Trouble In Town   
11                                      Versatile   
12                                       Крыльями   
13            Last Night in Sant Celoni - 12" Mix   
14                                       Medellín   
17                                          Пожар   
18                                          Insan   
19        The World Retreats - Marino Canal Remix   
20                                      Opa Gäärd   
21                                          sugar   
22                                 A Child's Tale   
23                                     Love Again   
24             Take Some Time - Emancipator Remix   
25                                 Time for Space   
26                          Shuggie - Bonus Track   
27                                     Borboletas   
28    When The Interlude Sounds Like Falling Rain   
29                               Phía Sau Đôi Mắt   
30  Suite No. 2 for Jazz Orchestra: VII. Waltz II   
31                                         Carmen   
32                                        Zukunft   
33                                    Bye Felicia   
34                                 I Had A Vision   
35                                 Lavish Lullaby   
36           Silver Tongue Devil (feat. Shenseea)   
37               Mystery Lady (feat. Don Toliver)   
38                                That Other Girl   
39                                         Weiter   
40                                   The Long Run   
41                                   Love Tonight   
42               Lily Moon - Isaac Chambers Remix   
43                     На своём вайбе (feat. Гуф)   
44                                          Мадам   
45                                    Ночной рейс   
46                           Один день в сентябре   
47                    С любимыми не расставайтесь   
48                              Ты - моя нежность   
49                               Забываю обещания   

                                          album                track_id  \
0                                    MOON FLAME  7Aoos39pDhB7H1egvGvImZ   
2                                         Light  7oyfJcOR2i3whfEYMogQt1   
3                                    Sputnik II  3IRON3rRIf2WJwejIbaehd   
4                          Good Morning Vietnam  1wOONZvB0H8xJhzo0vU0k7   
5                               SUGAR LIKE SALT  4o5MsHEU7bPIzOo49mv92l   
6                                        Trauma  4VOuvOhiGrsPxlDXF3l2Y6   
7                           Things Are Changing  3b4WEsIhYW8DAEImZzd98C   
8                            From the Beginning  30tifs9tO3pIW3gc5HlsTI   
9                               Kaleidoscope EP  7HBnZdg7fIQwqMhQhci0VV   
10                                Everyday Life  45PqOIkZ9PdCjsCJQYzx9G   
11                                    Versatile  2I1W5Mb4w7CIh3TyddCxc2   
12                                     Источник  1wAkp6u1biHGKX2q5IwBhf   
13  Last Night in Sant Celoni (feat. Jaz James)  6zAqmi9YkgnMxbqLcf7hDv   
14                         #EnAttendantMachakil  4Nmrb5DL6lN5kkb7LzX8yE   
17                                   Расстояние  772bNQ8WjMAXwiBywGcHb7   
18                                        Insan  01ZAEc3eyGGJLLZuAgdSMD   
19      The World Retreats (Marino Canal Remix)  5dyVrDue2N9ArxW34JyeO4   
20                                   A Long Way  1eb0mORiTlz0OLkH0NPb9Z   
21                                 next chapter  5IjIbGO7lih9CVDBFLCt

In [44]:
sample_n = (25-len(new_playlist_df))//2
sample_n

3

In [45]:
new_playlist_df = pd.concat([new_playlist_df,filtered_saved_m.sample(sample_n),filtered_saved_t.sample(sample_n)])
new_playlist_df.reset_index(drop = True, inplace= True)

In [46]:
new_playlist_df

track_name                           album  \
0                                  RSVP                      Nimmerland   
1          Napauken - Jpattersson Remix                           Zehna   
2                              Bläulich                     Treppenhaus   
3                       ABER WO BIST DU                           NACHT   
4                            Until Dawn                      Until Dawn   
5                         Rollercoaster                         T.H.A.L   
6             Struggle Made Me Stronger         Headphones on World Off   
7                              Beyikrta                 Chelina, Vol. 1   
8                  X Gon' Give It To Ya                 The Best Of DMX   
9             Berlin lebt wie nie zuvor                   Berlin lebt 2   
10                                Creep                     Pablo Honey   
11                   IC3 (feat. Skepta)              IC3 (feat. Skepta)   
12                   FEIND (feat. Azad)                           NACHT   
13                          Lambo Lambo                     KitschKrieg   
14                 Who - Single Version                             Who   
15               AUGEN ZU (feat. Samra)                           NACHT   
16                                  505       Favourite Worst Nightmare   
17                               Wealth                        Who Else   
18                               M.I.A.                      Nimmerland   
19                         Five Minutes                    Five Minutes   
20       Risky Business - Mathame Remix  Risky Business (Mathame Remix)   
21                          Fall Slowly                     Fall Slowly   
22                                Мадам                           Мадам   
23  Good Morning Vietnam - Original Mix            Good Morning Vietnam   
24                               Carmen                     Born To Die   

                  track_id               artist  duration  popularity  \
0   38UlieEW0eto55PNH9Z0cK                  RIN    169293          57   
1   2vPP6FNAzYoo2pplHp7Vop               Shkoon    380616          50   
2   2WRTnY0slmFgWcrmEr8dPj           Apache 207    196213          71   
3   2dGtLJysTSI9cbQ6TulL8V                 ELIF    185274          56   
4   1R2rmmUufa1tdiD9j9Qs3x               Jaeger    287190          49   
5   2uWnoJg6wgaSfBoaLPuwfl           badmómzjay    165833          31   
6   3vMtLNMANvkPuDthcLuQzJ  Fearless Motivation    213000          48   
7   6Kd77EjLuU4b8NZyOEkyTu              Chelina    216410           7   
8   2NeHnSFnwNp1Z5WYgcjJ8L                  DMX    219253           0   
9   2pfio0uHpT4USTPBNeTIWo          Capital Bra    151294          59   
10  6b2oQwSGFkzsMtQruIWm2p            Radiohead    238640          69   
11  0BMszcKwCrUDHZX3CEEj6L               Ghetts    231106          58   
12  4Upt2Q8OcUd85uo5x9QqWK                 ELIF    169299          53   
13  7oqvRZNv4dUV8CgQWtIAMe          KitschKrieg    214991          59   
14  0CjBORMsmiQNe3vPDcNIvk         Modeselektor    207988          44   
15  6ikiAn9th3TUfS3bYS7gDX                 ELIF    174248          71   
16  0BxE4FqsDD1Ot4YuBXwAPp       Arctic Monkeys    253586          82   
17  5aOlYhQsp75cgPov4yjWIe         Modeselektor    247218          43   
18  06pbKFx2Iut0MRs1XjV9Mc                  RIN    190640          50   
19  3pJnvBbwIvuwebHHTYzoR3                  Her    224672           0   
20  5R4hprpCcdgKz1DsPoh9p2                  ZHU    410078          44   
21  01WOwxkxOw2FqNIHkraxcN         Joyner Lucas    208095          63   
22  0keCs8wD67bWkE4yrqZmir                 JONY    171000          61   
23  1wOONZvB0H8xJhzo0vU0k7        Adrien Kepler    506666          40   
24  30yDbkiLvSiYx3pox7odMR         Lana Del Rey    248720          42   

    danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.787   0.671    9    -4.447     1       0.0409      0.064200   
1          0.793   0.294   

### Adding new tracks from Spotify recommendations

In this last step I add new tracks to fill up the other half of the playlist.

I **don't** want to simply add songs spotify recommends based on the songs, which are already in the playlist.

Therefore getting Spotify recommendations is only the first step. I am retrieving multiple songs recommendations per song, which are then filtered again based on similarity scoring.

In [47]:
seed_tracks = new_playlist_df["track_id"].tolist()
#seed_artists = artists_m["name"].tolist() + artists_t["name"].tolist()

In [48]:
len(seed_tracks)

25

Unfortunately **the Spotify API does not accept 25 seed tracks for a recommendation query**, I therefore am splitting up the process into "packages" of 5 seed tracks, retrieving 25 tracks per "package"

In [49]:
seed_tracks[:5], seed_tracks[5:10], seed_tracks[10:15]

(['38UlieEW0eto55PNH9Z0cK',
  '2vPP6FNAzYoo2pplHp7Vop',
  '2WRTnY0slmFgWcrmEr8dPj',
  '2dGtLJysTSI9cbQ6TulL8V',
  '1R2rmmUufa1tdiD9j9Qs3x'],
 ['2uWnoJg6wgaSfBoaLPuwfl',
  '3vMtLNMANvkPuDthcLuQzJ',
  '6Kd77EjLuU4b8NZyOEkyTu',
  '2NeHnSFnwNp1Z5WYgcjJ8L',
  '2pfio0uHpT4USTPBNeTIWo'],
 ['6b2oQwSGFkzsMtQruIWm2p',
  '0BMszcKwCrUDHZX3CEEj6L',
  '4Upt2Q8OcUd85uo5x9QqWK',
  '7oqvRZNv4dUV8CgQWtIAMe',
  '0CjBORMsmiQNe3vPDcNIvk'])

In [50]:
recomms = sp_m.recommendations(seed_tracks = seed_tracks[:5],limit = 25)

In [51]:
append_audio_features(create_df_recommendations(recomms),sp_m)

track_name                               album  \
0                       Mile Away                           Mile Away   
1                      Shot Clock                        Flyest Alive   
2                MASTERS VOM MARS  OUTTATHISWORLD - RADIO SHOW VOL. 1   
3                      Mehr davon                          Mehr davon   
4                           Chaos                          Armageddon   
5   Tarlabasi - Be Svendsen Remix                   Indoor Voyager EP   
6                          Perkys               Geld Motivierte Muzik   
7                 Thinking Bout U                     Thinking Bout U   
8                        Untitled                          Halcyon EP   
9                   Viel leichter                       Viel leichter   
10    Return to Oz - ARTBAT Remix         Return to Oz (ARTBAT Remix)   
11        Bap U - Not Sorry Remix             Bap U (Not Sorry Remix)   
12              Allein gegen Alle                       Frost Forever   
13                          Panik                          Paranoid!?   
14                    Monte Carlo                         Monte Carlo   
15                 Play With Fire                         Nico Santos   
16        stoned & allein - Remix             stoned & allein (Remix)   
17                           Baga                                Baga   
18                          Dough                               Dough   
19                            XXL                                 XXL   
20                  Smile For You                         Wild Window   
21                    Mitternacht                         Mitternacht   
22                         Donuts                                Gzuz   
23                          Krank                               Weiss   
24             Revolution - Remix                  Revolution (Remix)   

                  track_id              artist  duration  popularity  \
0   4Fl3rGS5f8s9QZ0otLX11N         Kelvyn Colt    193548          61   
1   0I4vdvCJcIoo0h3rEU39jA               Elias    150935          51   
2   0yBCdH5ZDUB55hMct48w4T            Genetikk    160100          41   
3   7rwAzZpANJEN0vn0pzYONW               LOTTE    173808          65   
4   4xUgmBAib7GP3U3Eg169Dg               Snavs    225000          32   
5   4cIDajUp2dMjKiDmDvFDg8  Oceanvs Orientalis    545813          58   
6   2qBl6nPXhJMTN7A5R9KvqM           Money Boy    134586          58   
7   1s3rLYo5NxTv7yQ0VZXxTl        Fabian Mazur    220800          39   
8   6Ni0ZItxBirgU2NCTmVaCK                AVEM    447693          36   
9   1Q7xCxhz6iwQNpnAUfYULZ                LUNA    208529          59   
10  0gwNGHcBRYtF7mvgUczVo1            Monolink    480000          67   
11  0ze0oWr0DVCCKupIX0RKE4         Party Favor    240172          32   
12  2ohNJM8og2aYwvX0DFsRGF              LGoony    214000          41   
13  0l1jmyo4eJ5D2UrMOFc8Ec         Mike Singer    149499          51   
14  2uuXKTgNVT6KU6w5aTpW3A           Money Boy    192527          60   
15  4XqCiPFWwAwGEI7yiq0LvK         Nico Santos    213933          68   
16  21cUGpkQZONi4EFowJqWUZ              Selmon    179620          59   
17  5oBkr3Zd8693UsdGqghZvH              Pronto    191356          55   
18  57n3WMBd7FK2WJNvvInlF2              FelMax    246015          39   
19  0p3euGPg5LgFeQGC7bBFuR     Miksu / Macloud    166173          72   
20  0FIJscQHEkL1c1tPhJUwoW       Mees Dierdorp    471774          48   
21  3PgxKVQx2ekpZQM7WhePiI         Vanessa Mai    183960          59   
22  5XVPNr0CAoUn1jD033Ucoo                Gzuz    162937          69   
23  2UFGOeae9LzuBgN1nmWwoV                Hava    182637          50   
24  7JAGL0wjf2lHoPUJ2oZF7w           Sean&Bobo    235153          56   

    danceability  energy  key  loudness  mode  speechiness  acousticness  \
0          0.548   0.553    7    -8.534     0       0.1650      0.170000   
1          0.780   0.572    8    -3.821     1       0.4570      0.019700   
2          0.468   0.87

In [52]:
recomm_dfs = []
for i in range(5,26,5):
    recomms = sp_m.recommendations(seed_tracks = seed_tracks[i-5:i],limit = 25)
    recomms_df = append_audio_features(create_df_recommendations(recomms),sp_m)
    recomm_dfs.append(recomms_df)
recomms_df = pd.concat(recomm_dfs)

In [53]:
recomms_df.reset_index(drop = True, inplace= True)

In [54]:
recomms_df

track_name  \
0                           Bleib mit mir wach   
1                         Never Gonna Catch Me   
2                                       TurnUp   
3                                     Aquafina   
4                                       CASINO   
..                                         ...   
120                          Time (feat. Rhye)   
121          Easier (feat. LOWES) - Radio Edit   
122                              Мне всё Монро   
123  Pinned to the Cross (feat. Finn Matthews)   
124                              Mother Africa   

                                         album                track_id  \
0                           Bleib mit mir wach  1qTyTDGi8O9rgDk3mheeuX   
1                         Never Gonna Catch Me  3folt4d0CndACKo02YNrin   
2                               All Trap Music  20Xwgtz25c8XBV46z1GYs4   
3                                 Flyest Alive  05fEXJyvQ7rji2J5Iri1Ix   
4                                   BOSS BITCH  5rywC2PH49hTpJLXxwGlpE   
..                                         ...                     ...   
120                          Time (feat. Rhye)  0XQypgwTXf0LS7ZGx13XTA   
121          Easier (feat. LOWES) [Radio Edit]  133YafRPaKLYCR28FSbN4M   
122                                         58  01LoBeQWux0KMNfiaqUFWp   
123  Pinned to the Cross (feat. Finn Matthews)  15hcBnrMPvMt8dBS54MDMS   
124                              Mother Africa  02uwXnLbfT68JKAShQ0VTS   

              artist  duration  popularity  danceability  energy  key  \
0               Xavi    150072          52         0.587   0.486    0   
1         El Speaker    200969          47         0.661   0.739    9   
2       Gent & Jawns    185753          39         0.603   0.972   11   
3              Elias    140869          50         0.862   0.851    1   
4    Katja Krasavice    201548          64         0.794   0.489    8   
..               ...       ...         ...           ...     ...  ...   
120         SG Lewis    259173          72         0.750   0.819   10   
121        CamelPhat    214687          66         0.471   0.895   11   
122       Egor Kreed    139742          54         0.820   0.617    2   
123        Rick Ross    261245          52         0.537   0.826    6   
124          TroyBoi    238181          55         0.739   0.690   10   

     loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
0      -8.902     0       0.0807       0.60400          0.000810    0.0976   
1      -5.221     0       0.0314       0.05290          0.064300    0.0559   
2      -5.013     0       0.0415       0.00205          0.687000    0.1240   
3      -1.473     1       0.0560       0.04550          0.000000    0.1450   
4      -8.341     0       0.0840       0.27900          0.000004    0.0966   
..        ...   ...          ...           ...               ...       ...   
120    -6.600     0       0.1510       0.06740          0.000979    0.3080   
121    -6.622     0       0.0458       0.04350          0.007470    0.1050   
122    -4.120     1       0.1510       0.24900          0.000000    0.0824   
123    -3.089     0       0.1940       0.10100          0.000000    0.2660   
124    -7.448     0       0.0314       0.14200          0.610000    0.1340   

     valence    tempo  
0      0.848  132.000  
1      0.550   98.003  
2      0.388  145.997  
3      0.541   91.945  
4      0.542  155.945  
..       ...      ...  
120    0.614  116.054  
121    0.114  122.954  
122    0.428  125.002  
123    0.437   76.997  
124    0.625  131.977  

[125 rows x 17 columns]

The 125 recommendations are further filtered by their similarity to the known tracks in the playlist.

In [55]:
similarity_score = create_similarity_score(new_playlist_df,recomms_df)

In [56]:
new_playlist_df.shape

(25, 17)

In [57]:
recomms_df

track_name  \
0                           Bleib mit mir wach   
1                         Never Gonna Catch Me   
2                                       TurnUp   
3                                     Aquafina   
4                                       CASINO   
..                                         ...   
120                          Time (feat. Rhye)   
121          Easier (feat. LOWES) - Radio Edit   
122                              Мне всё Монро   
123  Pinned to the Cross (feat. Finn Matthews)   
124                              Mother Africa   

                                         album                track_id  \
0                           Bleib mit mir wach  1qTyTDGi8O9rgDk3mheeuX   
1                         Never Gonna Catch Me  3folt4d0CndACKo02YNrin   
2                               All Trap Music  20Xwgtz25c8XBV46z1GYs4   
3                                 Flyest Alive  05fEXJyvQ7rji2J5Iri1Ix   
4                                   BOSS BITCH  5rywC2PH49hTpJLXxwGlpE   
..                                         ...                     ...   
120                          Time (feat. Rhye)  0XQypgwTXf0LS7ZGx13XTA   
121          Easier (feat. LOWES) [Radio Edit]  133YafRPaKLYCR28FSbN4M   
122                                         58  01LoBeQWux0KMNfiaqUFWp   
123  Pinned to the Cross (feat. Finn Matthews)  15hcBnrMPvMt8dBS54MDMS   
124                              Mother Africa  02uwXnLbfT68JKAShQ0VTS   

              artist  duration  popularity  danceability  energy  key  \
0               Xavi    150072          52         0.587   0.486    0   
1         El Speaker    200969          47         0.661   0.739    9   
2       Gent & Jawns    185753          39         0.603   0.972   11   
3              Elias    140869          50         0.862   0.851    1   
4    Katja Krasavice    201548          64         0.794   0.489    8   
..               ...       ...         ...           ...     ...  ...   
120         SG Lewis    259173          72         0.750   0.819   10   
121        CamelPhat    214687          66         0.471   0.895   11   
122       Egor Kreed    139742          54         0.820   0.617    2   
123        Rick Ross    261245          52         0.537   0.826    6   
124          TroyBoi    238181          55         0.739   0.690   10   

     loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
0      -8.902     0       0.0807       0.60400          0.000810    0.0976   
1      -5.221     0       0.0314       0.05290          0.064300    0.0559   
2      -5.013     0       0.0415       0.00205          0.687000    0.1240   
3      -1.473     1       0.0560       0.04550          0.000000    0.1450   
4      -8.341     0       0.0840       0.27900          0.000004    0.0966   
..        ...   ...          ...           ...               ...       ...   
120    -6.600     0       0.1510       0.06740          0.000979    0.3080   
121    -6.622     0       0.0458       0.04350          0.007470    0.1050   
122    -4.120     1       0.1510       0.24900          0.000000    0.0824   
123    -3.089     0       0.1940       0.10100          0.000000    0.2660   
124    -7.448     0       0.0314       0.14200          0.610000    0.1340   

     valence    tempo  
0      0.848  132.000  
1      0.550   98.003  
2      0.388  145.997  
3      0.541   91.945  
4      0.542  155.945  
..       ...      ...  
120    0.614  116.054  
121    0.114  122.954  
122    0.428  125.002  
123    0.437   76.997  
124    0.625  131.977  

[125 rows x 17 columns]

In [58]:
[np.argmax(i) for i in similarity_score]

[45,
 7,
 105,
 45,
 111,
 0,
 102,
 47,
 42,
 71,
 40,
 26,
 45,
 13,
 104,
 71,
 121,
 47,
 73,
 30,
 108,
 98,
 20,
 116,
 78]

In [59]:
final_recomms=recomms_df.loc[[np.argmax(i) for i in similarity_score]]
final_recomms = final_recomms.drop_duplicates()

In [60]:
new_playlist_df = new_playlist_df.append(final_recomms)

In [61]:
new_playlist_df = new_playlist_df.drop_duplicates()
new_playlist_df.reset_index(drop = True, inplace = True)

# The playlist is finished!

Now the only thing left to do is to add the tracks to the playlist.

(Adding a nice picture and thanking your girlfriend for her patience in the playlist description are **not optional**)

In [62]:
new_playlist_df

track_name  \
0                                   RSVP   
1           Napauken - Jpattersson Remix   
2                               Bläulich   
3                        ABER WO BIST DU   
4                             Until Dawn   
5                          Rollercoaster   
6              Struggle Made Me Stronger   
7                               Beyikrta   
8                   X Gon' Give It To Ya   
9              Berlin lebt wie nie zuvor   
10                                 Creep   
11                    IC3 (feat. Skepta)   
12                    FEIND (feat. Azad)   
13                           Lambo Lambo   
14                  Who - Single Version   
15                AUGEN ZU (feat. Samra)   
16                                   505   
17                                Wealth   
18                                M.I.A.   
19                          Five Minutes   
20        Risky Business - Mathame Remix   
21                           Fall Slowly   
22                                 Мадам   
23   Good Morning Vietnam - Original Mix   
24                                Carmen   
25                          Konjo Mewded   
26                             Aqua Drop   
27  What's On Your Mind (feat. Jacquees)   
28                                 NOPUS   
29                    Bleib mit mir wach   
30                        Always Forever   
31                    We Want Some P--sy   
32                                  Fiji   
33                           Bruce Wayne   
34                                LaLaLa   
35                      Dawn of an Alpha   
36                              Illusion   
37                             Riot Gear   
38     Easier (feat. LOWES) - Radio Edit   
39                               Überall   
40                        Tefa Yemileyen   
41       Breathing - Boris Brejcha Remix   
42                Berlin lebt immer noch   
43                      Tränen aus Kajal   
44                     Lys (feat. Menke)   
45                              Atlantis   

                                   album                track_id  \
0                             Nimmerland  38UlieEW0eto55PNH9Z0cK   
1                                  Zehna  2vPP6FNAzYoo2pplHp7Vop   
2                            Treppenhaus  2WRTnY0slmFgWcrmEr8dPj   
3                                  NACHT  2dGtLJysTSI9cbQ6TulL8V   
4                             Until Dawn  1R2rmmUufa1tdiD9j9Qs3x   
5                                T.H.A.L  2uWnoJg6wgaSfBoaLPuwfl   
6                Headphones on World Off  3vMtLNMANvkPuDthcLuQzJ   
7                        Chelina, Vol. 1  6Kd77EjLuU4b8NZyOEkyTu   
8                        The Best Of DMX  2NeHnSFnwNp1Z5WYgcjJ8L   
9                          Berlin lebt 2  2pfio0uHpT4USTPBNeTIWo   
10                           Pablo Honey  6b2oQwSGFkzsMtQruIWm2p   
11                    IC3 (feat. Skepta)  0BMszcKwCrUDHZX3CEEj6L   
12                                 NACHT  4Upt2Q8OcUd85uo5x9QqWK   
13                           KitschKrieg  7oqvRZNv4dUV8CgQWtIAMe   
14                                   Who  0CjBORMsmiQNe3vPDcNIvk   
15                                 NACHT  6ikiAn9th3TUfS3bYS7gDX   
16             Favourite Worst Nightmare  0BxE4FqsDD1Ot4YuBXwAPp   
17                              Who Else  5aOlYhQsp75cgPov4yjWIe   
18                            Nimmerland  06pbKFx2Iut0MRs1XjV9Mc   
19                          Five Minutes  3pJnvBbwIvuwebHHTYzoR3   
20        Risky Business (Mathame Remix)  5R4hprpCcdgKz1DsPoh9p2   
21                           Fall Slowly  01WOwxkxOw2FqNIHkraxcN   
22                                 Мадам  0keCs8wD67bWkE4yrqZmir   
23                  Good Morning Vietnam  1wOONZvB0H8xJhzo0vU0k7   
24                           Born To Die  30yDbkiLvSiYx3pox7odMR   
25            Evangadi Production, Vol 1  5CSSjkblq4KEome7dSztB9   
26                             Aqua Drop  3IR7oFtwy29YHudz2kK9Co   
27                                Kiss 5  0te0mYcVMqRLc4v

In [63]:
new_playlist_df["track_id"].tolist()

['38UlieEW0eto55PNH9Z0cK',
 '2vPP6FNAzYoo2pplHp7Vop',
 '2WRTnY0slmFgWcrmEr8dPj',
 '2dGtLJysTSI9cbQ6TulL8V',
 '1R2rmmUufa1tdiD9j9Qs3x',
 '2uWnoJg6wgaSfBoaLPuwfl',
 '3vMtLNMANvkPuDthcLuQzJ',
 '6Kd77EjLuU4b8NZyOEkyTu',
 '2NeHnSFnwNp1Z5WYgcjJ8L',
 '2pfio0uHpT4USTPBNeTIWo',
 '6b2oQwSGFkzsMtQruIWm2p',
 '0BMszcKwCrUDHZX3CEEj6L',
 '4Upt2Q8OcUd85uo5x9QqWK',
 '7oqvRZNv4dUV8CgQWtIAMe',
 '0CjBORMsmiQNe3vPDcNIvk',
 '6ikiAn9th3TUfS3bYS7gDX',
 '0BxE4FqsDD1Ot4YuBXwAPp',
 '5aOlYhQsp75cgPov4yjWIe',
 '06pbKFx2Iut0MRs1XjV9Mc',
 '3pJnvBbwIvuwebHHTYzoR3',
 '5R4hprpCcdgKz1DsPoh9p2',
 '01WOwxkxOw2FqNIHkraxcN',
 '0keCs8wD67bWkE4yrqZmir',
 '1wOONZvB0H8xJhzo0vU0k7',
 '30yDbkiLvSiYx3pox7odMR',
 '5CSSjkblq4KEome7dSztB9',
 '3IR7oFtwy29YHudz2kK9Co',
 '0te0mYcVMqRLc4vuDjZ0Yg',
 '02dPa4nXABwnFzjZosKxsk',
 '1qTyTDGi8O9rgDk3mheeuX',
 '2enPRFda84VE2wtI8c86Uf',
 '62ZlScwF8VZuGEFaAo3TNZ',
 '0sqLBIxpehB3UWkLHktEyo',
 '3huNXh7TCbzi9DlqCzhrUS',
 '4RvS94Li3lohRtTSz5X3xZ',
 '5G5EhGZzuBIU7kvfzmjguL',
 '0VRiX7eu12rWKnhuef6sa7',
 

**Note:** Here I am using `user_playlist_add_tracks()` to **add** to an existing playlist. It is possible to create one from scratch, however it wasn't necessary here. 

In the script I am using `playlist_replace_items()` as I don't just want new songs to be added but old ones to be deleted.

In [64]:
sp_m.user_playlist_add_tracks(usernames[0],
                              playlist_id="spotify:playlist:1Vcqtv3nE7QOJ4KFvK7bT8",
                              tracks = new_playlist_df["track_id"].tolist())

{'snapshot_id': 'MjAsMDkyYjQ3ZWNiMTc5NTM3ZGQwMTA2MTFlZTcyYzczYTljMzA0NzZlZg=='}

In [65]:
new_playlist_df.to_csv(path/"Playlist.csv")